### ISSUE TITLE
dispatcher: Replace with SCMP Daemon

### ISSUE THREAD

**Issue by scrye** [scionproto/scion#3961]
The dispatcher is the cause of many issues:
- It is a single point of failure for the SCION network stack of a host. If it goes down, a lot of processes (control service, gateway, tooling) go haywire.
- It lives in userspace, moving every forwarded packet from kernelspace to userspace and back to kernelspace, meaning a performance hit.
- It requires a UNIX DOMAIN stream socket, meaning awkward volume mounts in dockerized environments (and awkward resource cleanup/access management).
- It imposes stateful connections that need to be reestablished by client applications if the dispatcher goes down and then up. This has been the cause of many bugs in the past, and the performance impact of the reconnection logic is currently unknown.
- The UNIX DOMAIN socket is a stream, while the data being sent/received are packets. This requires packetization logic and slows forwarding even further.
- It has cost us a lot of time in performance tuning, because we never knew if it is the root cause of performance problems in apps like the POSIX gateway.
- It is an additional application to configure/deploy/operate on which, crucially, other applications depend on. This has caused headaches in testing where apps need to retry to connect to the dispatcher if it is not up yet, and it led to dependencies in `docker-compose`.
- It masks what applications are actually talking on the network, so a simple packet capture is often not sufficient to debug.
- It breaks native network tooling; `netstat` is ~ useless due to it.

The dispatcher has the following advantages:
- All SCION traffic is tunneled through a single port; this simplifies connectivity matrices and port firewall rules.
- The last SCION router on a path does not need to look up the L4 header to detect the underlay port, because it can assume 30041 there.

Also, the dispatcher as a stand-alone application (and not a forwarder) implements the logic to reply to certain SCMP messages like SCMP ping.

### The proposal

We should remove the dispatcher and replace it with an SCMP Daemon that only handles SCMP traffic.

SCION applications should open a UDP underlay port directly on the local host, on the same port as the UDP/SCION port. This will make both the port and the IP address match on both underlay and SCION.

Note that this is not a SCION protocol breaking change, because this is AS specific.

The roadmap would look like the following:
- Add a feature flag to the posix router, changing its behavior to set the UDP destination underlay port to the value of the UDP destination SCION port _only when they are the last router on the path_. Crucially, _this does not affect the performance of the high-speed core routers that need to forward huge volumes of data_.
- Have SCION applications use native `net.PacketConn` instead of `reliable.Conn`. This only needs changes in `snet`, and because the interfaces are compatible it is trivial to do.
- Create a new application called the SCMP Daemon (`scmpd`) that replies to SCMP traffic that targets a host (`scion ping`), forwards SCMP errors to the local SCION apps UDP ports based on the quotes, and forwards SCMP ping/traceroute requests received from apps to the next hop router. SCMP `ping`/`traceroute` become gRPC requests to `scmpd`. Normal SCION UDP sockets lose the ability to send out `ping`/`traceroute` (just like in the Linux kernel).
- Remove `ReconnectToDispatcher` from everywhere.

Thoughts?

**Comment by matzf** [comment]
Very nice, thanks a lot for this detailed proposal! I quite like this!

One point that I'm not sure about is the idea to "rewrite" the UDP destination
port in the border router.
My, admittedly very vague, concern is that the packets on the wire will look very
odd, with two nested UDP packets that must contain the same destination port.
It feels like this would be quite hard to explain.
It also feels like this would be moving away from what I imagine is the target
situation, where SCION/UDP is in some form directly supported by the end hosts
operating system.

For this, I would like to propose the following alternative:

### Rewrite UDP destination port on end host

- the border router is unchanged, i.e. it keeps addressing packets to a fixed destination port
- applications use native UDP connections, like you described
- the dispatcher still exists. It is modified to forward incoming packets to
  the (local) destination UDP port by using the address/port contained in the
  SCION and SCION/UDP header.
  The dispatcher is a dumb packet forwarder here, without state. In particular,
  it does not to be available for applications to open sockets etc.
- as a strictly optional optimisation, create an operating system specific tool
  that can do the port rewriting very early in the networking stack, e.g. using
  XDP.
  This can be run in parallel with the dispatcher; packets that are handled
  here simply never reach the dispatcher.  Therefore, this can simply pass
  through SCMPs and anything complicated to the actual user-space dispatcher.

So, the only real difference to the original proposal is that the destination
port "rewrite" happens at the end host, instead of in the border router.

**Comment by rreive** [comment]
Sorry i will try not to beat her in the future

**Comment by matzf** [comment]
The dispatcher does not currently implement SCMP port unreachable (`SCMPCodePortUnreachable`) messages. If we would want to implement this, though, the original proposal where the border router sends packets directly to the UDP destination would seem to make this rather cumbersome. The router would have to handle ICMP port unreachable messages, and convert them to SCMPs -- this is not impossible of course, but it appears like undesirable additional complexity for the router.
With the approach of having a dispatcher as proposed above, this could be fully handled on the end host.

**Comment by mlegner** [comment]
_(Disclaimer: I'm not very familiar with the end-host stack, please excuse any misunderstandings.)_

Overall I think @matzf's suggestion is somewhat "cleaner" in the sense that SCION border routers don't need to look at the L4 payload of packets. I'm not really worried about how the packets look on the wire but we should avoid introducing new middleboxes that break the layering concept in my opinion.

By having a very stripped-down dispatcher, which essentially just looks at the L4 header and forwards the packet, many of the disadvantages that @scrye listed are solved or at least become less severe:

- While still being a single point of failure, a much simpler version of the dispatcher is likely more stable and can be analyzed more thoroughly.
- The very limited functionality can more easily be implemented as a kernel module or in XDP as @matzf suggests.
- Stream sockets are no longer required as far as I understand.
- As it is no longer stateful, restarts of the dispatcher barely affect applications.
- The impact on performance is more predictable as it only performs very simple actions.

I'm not quite sure about the last three points though...

**Comment by scrye** [comment]
@matzf @mlegner : Thanks for the feedback. While we don't have the time to make progress on the rework of the dispatcher at the moment, it is something that will happen in the following months (we're currently looking at somewhere around April to start working on this).

No designs are in place yet, so if you have more ideas on how this could be tackled, we would really like to hear them. The more data/ideas we collect before we start, the better.

**Comment by martenwallewein** [comment]
Hey all,

since we (small team at Otto-von-Guericke University, Magdeburg) struggled a lot with the dispatcher performance in the current SCION stack, so we started to prototype a BPF in an XDP application that does the L4 port rewriting before the packets arrive at the dispatcher. Therefor, we changed also `snet.listen` to listen on UDP ports instead over the dispatcher. The prototype itself works for a fixed range of ports and we managed to start some simple benchmarks.

From performance perspective, we reached an improvement of around 60% (from 5Gbit/s to 8Gbit/s) in a first basic benchmark. However, it seems that we can not send packets fast enough, but this is another issue. Consequently, there seems to be a lot space left for improvements.

From usability perspective, the approach feels quite good, we did not need to change anything in the router or the dispatcher, we only needed to deploy the XDP prototype and build the benchmark app with our adapted snet lib. Other apps or the control services are not touched at all.

Although this approach covers only a small part of this  issue, we would like to share further progress and impressions on this here if everyone is fine with this.

Best,
Marten

**Comment by Oncilla** [comment]
@martenwallewein great to hear that there is interest in this topic.

Yes, please share your progress and insights here as you see fit. 
It is much appreciated :+1:

**Comment by matzf** [comment]
> In a completely unrelated meeting to this, the idea came up if we could do something like this: The application sends only the data along with some SCION metadata (like target host, port, path etc) to an application that encapsulates it to SCION packets and in the other direction decapsulates SCION packets back to only data. This would cover the design idea of having all SCION traffic though the same port (so we do not have to implement this in the routers) and gives the option to implement this deeper inside the OS to reach higher performance. What do you think about this?

This sounds like it's mostly equivalent to the current dispatcher setup, only that currently the fully serialized SCION packet is sent to the dispatcher. Conceptually, it doesn't make much of difference whether you pass this metadata in the form of the SCION header or pass it separately from the payload in a different, bespoke format.

**Comment by martenwallewein** [comment]
Ah of course, you're absolutely right. Only thought about implementation, not the conceptual perspective, which is indeed quite the same. So lets forget this idea for now.

**Comment by matzf** [comment]
We've recently had a discussion about the challenges of packaging/installing SCION-aware applications for different platforms. Specifically, one of the things that we intend to tackle is shipping a webbrowser with SCION support to Linux, Windows, MacOS, iOS and Android. It should be possible to install and run multiple of such SCION-aware applications simultaneously, i.e. directly listening on the "SCION end host port" 30041 for individual application is not an option.
It became clear that it's very cumbersome to have a component like the dispatcher that would have to be shared by all SCION-aware applications running on a device. Many of the targeted platforms not really have a useful concept of  "package dependencies" and even starting up the central SCION "dispatcher" component would either have to involve user interaction or would otherwise be platform specific.

This proposal to get rid of the need for a dispatcher would thus drastically simplify packaging and running SCION aware applications.

One remaining smaller issue with the original proposal is the proposed per-AS feature flag to enable the special UDP port selection behavior of the router;
We envision that at some point, there will be SCION aware applications are installed on many, many end user devices. At some point, SCION support may be added to some operating system network stacks, obviating the need for this clumsy workaround. At this point, we'll then want to disable this workaround (and perhaps SCION packets will then be delivered to end hosts directly over IP instead over IP/UDP). We should have a strategy to gradually phase-out the workaround and avoid requiring support on _all_ platforms before being able to disable it.
Our proposal for this would be to activate the scion-udp-port-defines-underlay-udp-port behavior in the router only for a specific _port range_. End hosts / applications making use of SCION support in the OS kernel (or hosts explicitly running a dispatcher component) use ports outside of this range.

The netsec folks and I would be willing to work the design and implementation for this.

**Comment by martenwallewein** [comment]
Thanks for catching up with this Matthias! I'm happy to help with design and implementation, for the workaround as well as for later steps bringing SCION into OS kernels. I personally like the idea of the workaround for the specific port range for the near future.

**Comment by matzf** [comment]
No objections have been raised so far (since I've retracted my previous objections, that is). The proposal review process is still a bit vague, but I think it's fair to just move one step forward now: The the final comment period of 10 days for this proposal starts now. If no objections are raised during this period, this proposal will be **accepted**.

The PR for the detailed design document (#4280) will undergo a separate final comment period once this proposal has been accepted.
